### PR TITLE
feat(#121): mem_base

### PR DESCRIPTION
**Pull request by h1alexbel**
In this pull I've implemented `MemBase` with [rusqlite](https://github.com/rusqlite/rusqlite) to transfer Storage logic from `xml/storage.rs` to `db/mem_base.rs` in the future PRs.

related to #121
History:
- **feat(#73): xpath**
- **feat(#73): no print**
- **feat(#121): pg_base**
- **feat(#121): ghiqc 0.0.5**
- **feat(#121): pg off, mem_base**
- **feat(#121): initial_schema**
- **feat(#121): checks_jeff_github**
- **feat(#121): puzzle**
- **feat(#121): puzzle for migrations**
- **feat(#121): puzzle, clean for cargo**

---

## PR-Codex overview
This PR introduces database handling improvements and licensing information updates.

### Detailed summary
- Added `rusqlite` dependency for database handling
- Updated `Storage` struct with `Serialize` and `Deserialize` traits
- Modified `register_user` function to use `Storage` object
- Added licensing information to multiple files
- Implemented `MemBase` struct for in-memory database handling
- Added `initial_schema` module for database schema initialization

> The following files were skipped due to too many changes: `server/src/db/initial_schema.rs`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: server/src/objects/user.rs [hunk 1]
```diff
@@ -61,20 +65,31 @@ impl User {
...
+        // if self
+        //     .storage
+        //     .xpath(&format!(
+        //         "//github/users/user[username='{}']",
+        //         self.username
+        //     ))
+        //     .contains(&self.username)
+        // {
+        //
+        // };
```
Comment (l3r8yJ) [inline server/src/objects/user.rs:77]: commented code

Comment (l3r8yJ) [inline server/src/objects/user.rs:77]: commented code in repo

File: server/src/objects/user.rs [hunk 2]
```diff
@@ -61,20 +65,31 @@ impl User {
...
     use crate::objects::user::User;
+    use crate::xml::storage::Storage;
     use anyhow::Result;
     use hamcrest::{equal_to, is, HamcrestMatcher};
 
     #[test]
     fn returns_username() -> Result<()> {
         let expected = "jeff";
-        let jeff = User::new(String::from(expected));
+        let jeff = User::new(String::from(expected), Storage::new(None));
```
Comment (l3r8yJ) [inline server/src/objects/user.rs:92]: user have a storage, wrong hierarchy I suppose

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel I have a general question. Why we need `sql` database? I think if we're choosing in memory solution we can use just some list(or any other collection) of data structs. IMHO it would be much simpler then have some `sql` storage. No requests needed. Need to add/remove something –  just call `github#add_user`/`github#add_repo` and etc. If we need some history of changes we can build some state machine that will track all changes that was made by user and it would build report based on it's history

**Comment by h1alexbel** [comment]
> @h1alexbel I have a general question. Why we need `sql` database? I think if we're choosing in memory solution we can use just some list(or any other collection) of data structs. IMHO it would be much simpler then have some `sql` storage. No requests needed. Need to add/remove something – just call `github#add_user`/`github#add_repo` and etc. If we need some history of changes we can build some state machine that will track all changes that was made by user and it would build report based on it's history

@l3r8yJ I thought about that. We can create some data structures in memory and operate with them. In the end we should obtain a hierarchy of these structures.

**Comment by h1alexbel** [comment]
@l3r8yJ Let's close #121, and proceed in separate ticket. This pull can be drafted.

**Pull request closed without merging**